File: week2/omerSefaYilmaz/min-stack.js
var MinStack = function () {
    this.minStack = [];
    this.stack = [];
  };
  
  /**
   * @param {number} val
   * @return {void}
   */
  MinStack.prototype.push = function (val) {
    this.stack.push(val);
  
    if (this.minStack.length < 1 || val <= this.getMin()) {
      this.minStack.push(val);
    } else {
      this.minStack.push(this.getMin());
    }
  };
  
  /**
   * @return {void}
   */
  MinStack.prototype.pop = function () {
    this.stack.pop();
    this.minStack.pop();
  };
  
  /**
   * @return {number}
   */
  MinStack.prototype.top = function () {
    return this.stack[this.stack.length - 1];
  };
  
  /**
   * @return {number}
   */
  MinStack.prototype.getMin = function () {
    if (this.minStack.length > 0) {
      return this.minStack[this.minStack.length - 1];
    }
  };
  
  /**
   * Your MinStack object will be instantiated and called as such:
   * var obj = new MinStack()
   * obj.push(val)
   * obj.pop()
   * var param_3 = obj.top()
   * var param_4 = obj.getMin()
   */

File: week2/AhmetTahaSener/MinStack.cs
public class MinStack
{

    public Stack<int> stack;
    public Stack<int> minStack;

    public MinStack()
    {
        stack = new Stack<int>();
        minStack = new Stack<int>();
    }

    public void Push(int val)
    {
        stack.Push(val);

        if (minStack.Count == 0 || val <= minStack.Peek())
        {
            minStack.Push(val);
        }
    }

    public void Pop()
    {
        if (stack.Count > 0)
        {
            int poppedValue = stack.Pop();

            if (minStack.Count > 0 && poppedValue == minStack.Peek())
            {
                minStack.Pop();
            }
        }
    }

    public int Top()
    {
        if (stack.Count > 0)
        {
            return stack.Peek();
        }
        return 0;
    }

    public int GetMin()
    {
        if (minStack.Count > 0)
        {
            return minStack.Peek();
        }
        return 0;
    }
}


/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.Push(val);
 * obj.Pop();
 * int param_3 = obj.Top();
 * int param_4 = obj.GetMin();
 */